### DISCUSSION TITLE
Externals (yes, again)

### DISCUSSION THREAD

**Discussion by samiam95124** [samiam95124/Pascal-P6#153] [General]
# Let's discuss externals, and a plan

## And it works like that

At present, externals are implemented "naturally". This takes some explanation. Anything can be external to the present program or module via a uses or joins declaration. Pascal-P6 does not really care what happens after that. You need to satisfy the uses or joins declaration with a module, but that is so that it can get the actual format of the declarations contained within during compile time. Linking with the actual code contained in the module isn't in fact done by Pascal-p6 at all. Its done by gcc, which really passes the job off to "ld", the GNU linker.

From Pascal-P6's point of view, the external module can contain all of the code and definitions, or it could be a stripped down version that just contains declarations. Pascal-P6 doesn't care. In fact, when it sees the "private" keyword in the file, meaning that no further public declarations follow, it just stops reading the external module file[1].

The external part is that if you compile that module in another language, usually C, but it could be anything, then give that to the gcc linker, you get defacto external linkage. Since Pascal-P6 formats the call according to the AMD64 ABI, any language using that calling convention will work, including C[3].

The consequences of this are that any symbol in the external file is formatted as:

module.name

IE, if the module is:

```
module mymod;

procedure myproc; begin end;

.
```

The external name is:

mymod.myproc

## What external does

The external keyword (in Pascaline it is a keyword, in ISO 7185 it isn't) gives you three very nice effects, and one that I personally find questionable (but hey, that is me):

1. external replaces the body of the procedure or function.
2. The parameters to it are not checked for use or non-use.
3. Setting a function result is not checked for.
4. It can be used anywhere.

The meaning of this is that a declaration file that declares external C (or other language) functions won't have to fake empty functions. Thus that is the preferred external representation. What about #4 there? Well, that means you can use the external declaration in the same module it is used in, just like in C or other languages. Why do I find that questionable? Because it's not really the business of the using module where the external comes from, thus I prefer putting it in an external module[2].

## I thunk that works

All of this is how it works now. You can use this system as is to do external linkages by using so called "thunks" to interface to external code. Thunks are external code that translates the call to adapt to other conventions. The minimum thunk needed to call external code in C is conversion of the name to it's non-module format. So in the above example, it would convert mymod.myproc() to just myproc().

## Enter cexternal

All of the above can be used to interface to any code in any language. It really just depends on your ability to write thunks, which can get pretty creative. Take callbacks as an example. However, the purpose of Pascal-P6 (or "A purpose") is to make it is easy as possible to link to external C code. So I propose the "cexternal" keyword (this was also covered in a previous article). cexternal is used in place of external to give extra capabilities to the external function.

## cexternal abilities

The first and most obvious effect of cexternal is that it drops the module name when calling externals. Thus it can be used to directly interface to external code without thunks. To be more specific, cexternal will automatically generate thunks as required.

For the rest, I propose a new type, a new mode, and a new parameter treatment that applies only to cexternal.

First of all, let get the rules straight. Whatever we do to make it easy to interface with C or other languages, we will NOT I repeat NOT import any of the type insecurities of C or other language to Pascal-P6. No casts, no void types, etc. Other Pascals do this. I am specifically avoiding it.

### stringz

stringz is a special type that is only applicable to cexternal headers. On value parameters, it is compatible with the type string in Pascal-P6, and it means that the input array of characters, of any size, is copied to a buffer and terminated with a zero. The address of the resulting string is passed to external code, without a length.

On out parameters, the type is compatible with container arrays of string only (arrays with length). On return, the zero terminated string has its length found, and that is used to set the length of the container pointer.

There is no var mode for stringz. The string either is an in parameter (value) or is an out parameter. Not both.

### ptrarr

ptrarr is a mode, and expects a pointer type for the parameter. All of the types that are compatible with the parameter pointer type are valid, including null, but in addition, any array which has the base type of the pointer are also compatible. The effect is that a pointer to the first element of the array is passed. Need I say this is a dangerous mode? This essentially treats ptrarr modes as a type of overload.

### procedure and function parameters

Finally, any procedure or function parameter gets special treatment in cexternal. C does not have function or procedure parameters, but rather it has function pointers. Since C, before C99, does not support nested functions, that is not a problem. In C99 it is, and C programmers handle it by never specifying a callback to a nested function. Do that, and all time and space collapses, clocks run backwards, all bills come due yesterday, etc. Pascal-P6 handles that by constructing a special kind of thunk that goes onto the stack. It passes a function pointer to C/Whatever, but that pointer really goes to the thunk on the stack, which is code that restores both the address of the actual callback function as well as its frame pointer. Thus the callback is valid, but not necessarily safe.

Why isn't it safe? Simple. The nested function or procedure could have exited in the meantime. If you call that, the code will certainly crash. The thunk onstack won't exist anymore.

## Mixed thunks

So what if you want to implement an external module that sends some functions to regular C code and some that implement thunks? Yea, that is possible too. In the external implementation file, you specify the direct calls as cexternal, and the thunked calls as external. The cexternals will go to regular C style names, and the externals will go to module.name type calls. Thus you can implement thunks just for the external flagged functions, and not the cexternal ones[4].

## What, *none* of that makes you happy? ch2ph

In constructing large files for external definitions, I encountered the issue before that the work becomes unreasonable fast. Like, for example, the windows.h header file, which contains most of the entire Windows API. The answer to that was ch2ph, which is a program from IP Pascal. This is the top end of an entire C compiler, including the macro preprocessor, married with a header file generator and thunk generator. ch2ph generates a header file for Pascal, by default in what I call "normal" format. The problem with translating C to Pascal is that certain constructs in C are ambiguous, and don't have direct equivalents in Pascal. The "normal" form means that ch2ph does a default translation with the assumption that you will fix it later. ch2ph also has a markup language to perform translations on the output headers, and features a macro language to describe how to create thunks.

ch2ph was used to translate the entire windows.h header file. It works because most of the C calls directly translate to Pascal calls, and then a database is used to describe the changes needed, and/or thunks needed for the exception cases. The goal achieved there was to not require modification of the input header file. In this way, any new header files that come from the writers of the external source can just get translated without change[5].

## Defacto type casting

So I reserved the ugliest subject for last. Since there is no way for Pascal-P6 to check types from the external module, the call to the external (non-Pascaline) function is completely blind. If you did, or didn't, get the parameter types right is up to you. This also means that you can use it to cast types. This practice is deeply evil, should never, ever be done, and of course I have done it myself several times.

[1] In fact, it is perfectly valid for a compiler to use a digested form of the external module, for example it's intermediate, to satisfy the external types.
[2] Just as it is bad practice in C as well.
[3] And yes, any language that obeys the AMD64 ABI calling convention, including Python, Rust, whatever will do. In fact, if you are willing to write a thunk, it doesn't even matter if it obeys the calling convention. You supply the adaption code in the thunk.
[4] Ok, yes, you could thunk cexternal calls. You would have to handle the difference in names between the cexternal and the target routine, which can no longer be the same. I prefer to thunk externals for this reason.
[5] ch2ph also can take "attributes" or macros that evaluate to nothing, and use those to form instructions on how to build headers. However that breaks the rule that we want header input files in C to be left unchanged, so the method is not used on existing header files.

**Comment by samiam95124** [comment]
# Me??? (the rebuttal)

Yea, this is in a reply for a reason, because it is my opinion.

Virtually none of what is supported for C interface, or planned for support to C interface, I think qualifies as good programming. Zero terminated string APIs have become famous for causing buffer overruns and cracker attacks[1], especially for return buffers. Further, virtually no other language supports them. The right answer is to pass strings as data/length pairs. The C community has come to this conclusion, which is why things like strncmp() exist. Duality between arrays and pointers is a C feature, certainly, but it is the way the language works. Programmers know which kind of parameter they mean, array or pointer, and I don't think I have seen an instance where both a pointer and an array was meant. The language is ambiguous, the programmers aren't. Thus picking a Pascal meaning for the parameter of either an array or pointer, but not both, is the proper way to do this. Its only a problem when APIs are involved.

Finally, and I realize I am going to offend many here, callbacks are just a bad idea in general. Saving a function and then suddenly calling it much later from code that the application called really wrecks the idea of program flow. I realize that with callbacks being popular, it is impossible to construct software without using callbacks since the API being used demands them (cough ... GTK+), but the right thing to do is put to put C callbacks into C code and then call that code instead of calling back to Pascal code.

I dealt with callbacks in Petit-Ami, and, although I exposed callbacks to client programs, I used that in the example programs just once. That was for a general program abort, which was easier to handle that way.

In short, I have left these decisions to the user, but that does not mean I don't have strong opinions on the matters at hand.

[1] Note please I said "in APIs".

**Comment by samiam95124** [comment]
# Expansions on ptrarr

Probably with the benefit of hindsight, the most impactful of the cexternal adaptions is the ptrarr mode. Why? Because pointer parameters in C very ambiguous. To review, they can actually mean three different things, and in the case of Windows, four:

void apifunc(int* x);

x can be any of:

1. A pointer to an integer;
2. An array x[].
3. A parameter that does not exist (NULL).
4. A small number (vs. the big numbers that are generally pointers).

That last one was encountered in the Windows API, and I think frankly is a sign of mental illness. Windows programmers noticed that anything less than 4096 (4k) was impossible, due to the first page in Windows memory being a guard page against stray pointers. Don't do that. Don't even think that. Anyways, ptrarr mode does not support it.

At first glance it would seem that differentiating between the first two cases, and thus the third one by association, would be easy in C. There is int* x and int x[], right? A NULL is then automatically a pointer. End of problem, we can go home now.

No such luck! Here is the official man page of strlen():

size_t strlen(const char *s);

Yea, its a pointer, not an array. C treats these forms as equivalent, and so do most programmers. This also means the NULL case is ambiguous.

In Pascal, of course, these forms are not equivalent:

type intp = ^integer;
        inta = array of integer;

procedure 

procedure apifunc(x: intp);

procedure apifunc(x: inta);

or expressed as an overload group:

procedure apifunc(x: intp);
overload procedure apifunc(x: inta);

Which was exactly how they were specified in IP Pascal and ch2ph. For each such function, the ch2ph control file needed to specify conversion to an overload. This was a lot of work, since the Windows.h API is quite large.

Of course, in the case of IP Pascal, it's a case of fitting a standard compiler to C. Thus, there had to be a different overload group generated for each type of parameter. In Pascal-P6, we can do anything we want, including some sketchy type fitting. Indeed, one option is to just take any array or pointer parameter and let either arrays or pointers fit that parameter, effectively automatically casting the type. And that may be the winning answer.

Is it dangerous? Of course. But this is true of any call to C. If you specified that a ham sandwich exists on the other side of a call to C, and it isn't so, a crash results. You might take comfort to the fact that the crash will occur on the C language side, if that helps.

**Comment by samiam95124** [comment]
# Expansions on stringz

## Value parameters

procedure x(s: stringz);

This is a value parameter s. var, or out is invalid because we are changing the string, that is, changing its length and adding the zero termination. stringz can be match any length of string, because it can create any length of string for the C call. If the call is:

var s: pstring;

x(s^);

Passes an address/length pair, but the length is used to size the buffer used to copy the string to, and then  not passed to the C program.

Implied in this is that the string gets buffered. There is no real way to pass the string in place.

With input or value parameters, the size of the buffer is controlled. on the Pascaline side, but of course not on the C side. However, this is the most type safe mode.

## Out parameters

procedure x(out s: stringz);

This does not work for var parameters, for the same reason that value parameters must be used for input parameters. The address of the string is passed to C, but the length is not. The C side fills out the buffer provided, including with a zero termination. On return to the Pascaline side, the zero termination is found, and the zero byte, and all further characters until the end of the buffer, are filled with spaces.

There's a lot that is insecure with this mode. The C side has no idea what the size of the buffer is, it is just an assumption. This is why this form of call is deprecated in C. Most of the time when a return API is used in C, the length of the buffer is also passed, which is another issue and requires another type of solution.

The other factor is that even with containers, the size of the buffer provided for the return is fixed. We can't even pass the length of the string back to the caller. All we can do is pad out the rest of the buffer with spaces.

## Conclusion

I think the result of this is that string in parameters to C are fairly useful,. and there are a lot of zero terminated in parameters to C APIs. One takeaway is that such parameters are slower than an address/length pair, which for this and other reasons is preferable.

For the out parameter case, it works but should be avoided. The biggest issue on the Pascaline side is that the length of the string is lost unless you count padded strings. The Pascaline method would be to pass a container pointer and allocate a string from dynamic memory string so that it and the length can be filled out. This could be an additional mode, but it also has issues, since we don't know who big a buffer to allocate and send to the C routine.
